File: problems/2026/04/2026-04-06-leetcode-17-电话号码的字母组合/SolutionLC17.py
from typing import List, Set


class Solution:
    KEYS = ("", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz")

    def letter_combinations(self, digits: str) -> List[str]:
        """回溯；时间 O(4^k * k)，递归栈 O(k)。"""
        if not digits:
            return []
        res: List[str] = []
        path: List[str] = []

        def dfs(index: int) -> None:
            if index == len(digits):
                res.append("".join(path))
                return
            for ch in self.KEYS[int(digits[index])]:
                path.append(ch)
                dfs(index + 1)
                path.pop()

        dfs(0)
        return res


if __name__ == "__main__":
    solution = Solution()

    inputs = ["", "2", "23"]
    expected: List[Set[str]] = [
        set(),
        {"a", "b", "c"},
        {"ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"},
    ]

    for i, s in enumerate(inputs):
        result = solution.letter_combinations(s)
        got = set(result)
        ok = got == expected[i] and len(result) == len(got)
        print(f'调用: letter_combinations("{s}")')
        print(f"预期结果: {expected[i]}")
        print(f"实际结果: {got}")
        print(f"校验: {'OK' if ok else 'FAIL'}")
        print()

File: problems/2026/04/2026-04-06-leetcode-17-电话号码的字母组合/SolutionLC17.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * LeetCode 17：电话号码的字母组合（回溯 / DFS）
 */
public class SolutionLC17 {

    private static final String[] KEYS = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

    public List<String> letterCombinations(String digits) {
        List<String> res = new ArrayList<>();
        if (digits == null || digits.isEmpty()) {
            return res;
        }
        StringBuilder path = new StringBuilder();
        dfs(digits, 0, path, res);
        return res;
    }

    private void dfs(String digits, int index, StringBuilder path, List<String> res) {
        if (index == digits.length()) {
            res.add(path.toString());
            return;
        }
        int d = digits.charAt(index) - '0';
        String letters = KEYS[d];
        for (int i = 0; i < letters.length(); i++) {
            path.append(letters.charAt(i));
            dfs(digits, index + 1, path, res);
            path.deleteCharAt(path.length() - 1);
        }
    }

    /** 与预期做集合比对（题目允许任意顺序）。 */
    private static boolean sameAsExpected(List<String> got, Set<String> expected) {
        return got.size() == expected.size() && new HashSet<>(got).equals(expected);
    }

    public static void main(String[] args) {
        SolutionLC17 solution = new SolutionLC17();

        String[] inputs = {"", "2", "23"};
        List<Set<String>> expected = Arrays.asList(
                new HashSet<>(),
                new HashSet<>(Arrays.asList("a", "b", "c")),
                new HashSet<>(Arrays.asList(
                        "ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"))
        );

        for (int i = 0; i < inputs.length; i++) {
            List<String> result = solution.letterCombinations(inputs[i]);
            boolean ok = sameAsExpected(result, expected.get(i));
            System.out.println("调用: letterCombinations(\"" + inputs[i] + "\")");
            System.out.println("预期结果: " + expected.get(i));
            System.out.println("实际结果: " + new HashSet<>(result));
            System.out.println("校验: " + (ok ? "OK" : "FAIL"));
            System.out.println();
        }
    }
}
